src/desktop/usb.rs

```diff
@@ -0,0 +1,259 @@
+//!     let session = usb.create_session().await?;
+//!     if let Some(response) = usb.receive_device_events().await?.next().await {
+//!         let events = response.events();
+//!         for ev in events {
+//!              println!(
+//!                  "Received event: {} for device {}",
+//!                  ev.event_action(),
+//!                  ev.event_device_id()
+//!              );
+//!         }
+//!     }
+//!
+//!     Ok(())
+//! }
+//! ```
+
+use std::collections::HashMap;
+
+use futures_util::Stream;
+use serde::Deserialize;
+use zbus::zvariant::{
+    DeserializeDict, Fd, ObjectPath, OwnedFd, OwnedObjectPath, OwnedValue, SerializeDict, Type,
+    Value,
+};
+
+use super::Request;
+use crate::{
+    desktop::{HandleToken, Session, SessionPortal},
+    proxy::Proxy,
+    Error, WindowIdentifier,
+};
+
+#[derive(Debug, SerializeDict, Type, Default)]
+#[zvariant(signature = "dict")]
+struct CreateSessionOptions {
+    handle_token: HandleToken,
+    session_handle_token: HandleToken,
+}
+
+/// Options for the USB portal.
+#[derive(SerializeDict, Type, Debug, Default)]
+#[zvariant(signature = "dict")]
+pub struct UsbOptions {
+    handle_token: HandleToken,
+    session_handle_token: HandleToken,
+}
+
+/// USB device description
+#[derive(SerializeDict, DeserializeDict, Type, Debug, Default)]
+#[zvariant(signature = "dict")]
+pub struct UsbDevice {
+    parent: Option<String>,
+    readable: Option<bool>,
+    /// Device is writable. Default is false.
+    pub writable: Option<bool>,
+    /// The device node for the USB.
+    #[zvariant(rename = "device-file")]
+    pub device_file: Option<String>,
+    /// Device properties
+    pub properties: Option<HashMap<String, OwnedValue>>,
+}
+
+impl UsbDevice {
+    /// Return the vendor string for display.
+    pub fn vendor(&self) -> Option<String> {
+        self.properties.as_ref().and_then(|properties| {
+            properties
+                .get("ID_VENDOR_FROM_DATABASE")
+                .or_else(|| properties.get("ID_VENDOR_ENC"))
+                .and_then(|v| v.downcast_ref::<String>().ok())
+        })
+    }
+
+    /// Return the model string for display.
+    pub fn model(&self) -> Option<String> {
+        self.properties.as_ref().and_then(|properties| {
+            properties
+                .get("ID_MODEL_FROM_DATABASE")
+                .or_else(|| properties.get("ID_MODEL_ENC"))
+                .and_then(|v| v.downcast_ref::<String>().ok())
+        })
+    }
+}
+
+/// Device to acquire.
+#[derive(SerializeDict, Type, Debug, Default)]
+#[zvariant(signature = "dict")]
+pub struct AcquireDevice {
+    writable: bool,
+}
+
+/// Option for AcquireDevice call.
+#[derive(SerializeDict, Type, Debug, Default)]
+#[zvariant(signature = "dict")]
+pub struct AcquireOptions {
+    handle_token: HandleToken,
+}
+
+/// Finished device acquired.
+#[derive(DeserializeDict, Type, Debug, Default)]
+#[zvariant(signature = "dict")]
+pub struct AcquiredDevice {
```

Comment:
This seems like it should be an internal type that gets mapped into a `Result<OwnedFd, UsbError>`, as it is, it is not really Rust friendly.